File: Battleship/UserInterface.cpp
#include "UserInterface.hpp"
#include "GameManager.hpp"
#include <iostream>

UserInterface::UserInterface(GameManager* manager)
	: m_gameManager(manager)
{
}

void UserInterface::DisplayBoards(Player* player)
{
	std::cout << "========================================\n";
	std::cout << "         ТЕКУЩЕЕ СОСТОЯНИЕ\n";
	std::cout << "========================================\n";
	std::cout << "Игрок: " << player->GetName() << "\n\n";

	// Поле игрока - показываем корабли (forOwner = true)
	std::cout << "=== ВАШЕ ПОЛЕ ===\n";
	auto myState = player->GetMyBoard().GetVisibleState(true);

	// Вывод номеров столбцов
	std::cout << "  ";
	for (int j = 0; j < player->GetMyBoard().GetSize(); j++)
	{
		std::cout << j << " ";
	}
	std::cout << "\n";

	for (int i = 0; i < player->GetMyBoard().GetSize(); i++)
	{
		std::cout << i << " ";
		for (int j = 0; j < player->GetMyBoard().GetSize(); j++)
		{
			std::cout << myState[i][j] << " ";
		}
		std::cout << "\n";
	}

	// Поле противника - НЕ показываем корабли (forOwner = false)
	std::cout << "\n=== ПОЛЕ ПРОТИВНИКА ===\n";
	auto enemyState = player->GetEnemyBoard()->GetVisibleState(false);

	// Вывод номеров столбцов
	std::cout << "  ";
	for (int j = 0; j < player->GetEnemyBoard()->GetSize(); j++)
	{
		std::cout << j << " ";
	}
	std::cout << "\n";

	for (int i = 0; i < player->GetEnemyBoard()->GetSize(); i++)
	{
		std::cout << i << " ";
		for (int j = 0; j < player->GetEnemyBoard()->GetSize(); j++)
		{
			std::cout << enemyState[i][j] << " ";
		}
		std::cout << "\n";
	}

	// Легенда
	DisplayLegend();
	std::cout << "========================================\n";
}

void UserInterface::DisplayMessage(const std::string& message)
{
	std::cout << message << "\n";
}

void UserInterface::ShowGameOver(const std::string& winnerName)
{
	std::cout << "\n========================================\n";
	std::cout << "           ИГРА ОКОНЧЕНА!\n";
	std::cout << "========================================\n";
	std::cout << "*** " << winnerName << " ПОБЕДИЛ В ИГРЕ! ***\n";

	// Показываем все корабли противника в конце игры
	if (m_gameManager)
	{
		Player* currentPlayer = m_gameManager->GetCurrentPlayer();
		if (currentPlayer)
		{
			std::cout << "\n=== РАСКРЫТОЕ ПОЛЕ ПРОТИВНИКА ===\n";

			// Используем forOwner = true чтобы показать все корабли противника
			auto revealedState = currentPlayer->GetEnemyBoard()->GetVisibleState(true);

			// Вывод номеров столбцов
			std::cout << "  ";
			for (int j = 0; j < currentPlayer->GetEnemyBoard()->GetSize(); j++)
			{
				std::cout << j << " ";
			}
			std::cout << "\n";

			for (int i = 0; i < currentPlayer->GetEnemyBoard()->GetSize(); i++)
			{
				std::cout << i << " ";
				for (int j = 0; j < currentPlayer->GetEnemyBoard()->GetSize(); j++)
				{
					std::cout << revealedState[i][j] << " ";
				}
				std::cout << "\n";
			}
		}
	}
	std::cout << "========================================\n";
}

void UserInterface::DisplayLegend()
{
	std::cout << "\n--- ЛЕГЕНДА ---\n";
	std::cout << "S - ваш корабль\n";
	std::cout << "X - попадание\n";
	std::cout << "O - промах\n";
	std::cout << ". - неизвестная клетка\n";
	std::cout << "----------------\n";
}

File: BattleshipCS/UserInterface.cs
namespace BattleshipCS;
public class UserInterface
{
    private readonly GameManager gameManager;

    public UserInterface(GameManager manager)
    {
        gameManager = manager;
    }

    public void DisplayBoards(Player player)
    {
        Console.WriteLine("========================================");
        Console.WriteLine("         ТЕКУЩЕЕ СОСТОЯНИЕ");
        Console.WriteLine("========================================");
        Console.WriteLine($"Игрок: {player.Name}\n");

        // Поле игрока
        Console.WriteLine("=== ВАШЕ ПОЛЕ ===");
        var myState = player.MyBoard.GetVisibleState(true);

        // Вывод номеров столбцов
        Console.Write("  ");
        for (int j = 0; j < player.MyBoard.Size; j++)
        {
            Console.Write($"{j} ");
        }
        Console.WriteLine();

        for (int i = 0; i < player.MyBoard.Size; i++)
        {
            Console.Write($"{i} ");
            for (int j = 0; j < player.MyBoard.Size; j++)
            {
                Console.Write($"{myState[i, j]} ");
            }
            Console.WriteLine();
        }

        // Поле противника
        Console.WriteLine("\n=== ПОЛЕ ПРОТИВНИКА ===");
        var enemyState = player.EnemyBoard!.GetVisibleState(false);

        // Вывод номеров столбцов
        Console.Write("  ");
        for (int j = 0; j < player.EnemyBoard.Size; j++)
        {
            Console.Write($"{j} ");
        }
        Console.WriteLine();

        for (int i = 0; i < player.EnemyBoard.Size; i++)
        {
            Console.Write($"{i} ");
            for (int j = 0; j < player.EnemyBoard.Size; j++)
            {
                Console.Write($"{enemyState[i, j]} ");
            }
            Console.WriteLine();
        }

        // Легенда
        DisplayLegend();
        Console.WriteLine("========================================");
    }

    public void DisplayMessage(string message)
    {
        Console.WriteLine(message);
    }

    public void ShowGameOver(string winnerName)
    {
        Console.WriteLine("\n========================================");
        Console.WriteLine("           ИГРА ОКОНЧЕНА!");
        Console.WriteLine("========================================");
        Console.WriteLine($"*** {winnerName} ПОБЕДИЛ В ИГРЕ! ***");

        // Показываем все корабли противника в конце игры
        var currentPlayer = gameManager.CurrentPlayer;
        if (currentPlayer != null)
        {
            Console.WriteLine("\n=== РАСКРЫТОЕ ПОЛЕ ПРОТИВНИКА ===");

            // Используем forOwner = true чтобы показать все корабли противника
            var revealedState = currentPlayer.EnemyBoard!.GetVisibleState(true);

            // Вывод номеров столбцов
            Console.Write("  ");
            for (int j = 0; j < currentPlayer.EnemyBoard.Size; j++)
            {
                Console.Write($"{j} ");
            }
            Console.WriteLine();

            for (int i = 0; i < currentPlayer.EnemyBoard.Size; i++)
            {
                Console.Write($"{i} ");
                for (int j = 0; j < currentPlayer.EnemyBoard.Size; j++)
                {
                    Console.Write($"{revealedState[i, j]} ");
                }
                Console.WriteLine();
            }
        }
        Console.WriteLine("========================================");
    }

    private void DisplayLegend()
    {
        Console.WriteLine("\n--- ЛЕГЕНДА ---");
        Console.WriteLine("S - ваш корабль");
        Console.WriteLine("X - попадание");
        Console.WriteLine("O - промах");
        Console.WriteLine(". - неизвестная клетка");
        Console.WriteLine("----------------");
    }
}